src/qutip_qip/circuit/text_renderer.py

```diff
@@ -0,0 +1,497 @@
+        xskip = self._get_xskip(wire_list, layer)
+        self._adjust_layer(wire_list, xskip)
+
+        width = 4 * ceil(self.style.gate_pad) + 1
+        cross_conn = "─" * (width // 2) + "╳" + "─" * (width // 2)
+        bar_conn = " " * (width // 2) + "│" + " " * (width // 2)
+        mid_bar_conn = "─" * (width // 2) + "│" + "─" * (width // 2)
+        self._manage_layers(width, wire_list, layer, xskip)
+
+        for wire in wire_list:
+            if wire == wire_list[-1]:
+                self._render_strs["top_frame"][wire] += " " * len(bar_conn)
+                self._render_strs["mid_frame"][wire] += cross_conn
+                self._render_strs["bot_frame"][wire] += bar_conn
+            elif wire == wire_list[0]:
+                self._render_strs["top_frame"][wire] += bar_conn
+                self._render_strs["mid_frame"][wire] += cross_conn
+                self._render_strs["bot_frame"][wire] += " " * len(bar_conn)
+            else:
+                self._render_strs["top_frame"][wire] += bar_conn
+                self._render_strs["mid_frame"][wire] += mid_bar_conn
+                self._render_strs["bot_frame"][wire] += bar_conn
+
+    def layout(self):
+        """
+        Layout the circuit
+        """
+        self._add_wire_labels()
+
+        for gate in self._qc.gates:
+
+            if isinstance(gate, Gate):
+                gate_text = (
+                    gate.arg_label if gate.arg_label is not None else gate.name
+                )
+
+            # generate the parts, width and wire_list for the gates
+            if isinstance(gate, Measurement):
+                wire_list = list(range(gate.targets[0] + 1)) + list(
+                    range(
+                        gate.classical_store + self._qwires,
+                        self._qwires + self._cwires,
+                    )
+                )
+                parts, width = self._draw_measurement_gate(gate)
+            elif len(gate.targets) == 1 and gate.controls is None:
+                wire_list = gate.targets
+                parts, width = self._draw_singleq_gate(gate_text)
+            elif gate.name == "SWAP":
+                self._update_swap_gate(gate)
+                continue
+            else:
+                merged_wire = sorted(gate.targets + (gate.controls or []))
+                wire_list = list(range(merged_wire[0], merged_wire[-1] + 1))
+                parts, width = self._draw_multiq_gate(gate, gate_text)
+
+            # update the render strings for the gate
+            layer = max(len(self._layer_list[i]) for i in wire_list)
+            xskip = self._get_xskip(wire_list, layer)
+            self._manage_layers(width, wire_list, layer, xskip)
+
+            if isinstance(gate, Measurement):
+                self._adjust_layer(wire_list, xskip)
+                self._update_singleq(gate.targets, parts)
+                self._update_cbridge(gate, wire_list, xskip, width)
```

Comment:
For clarity
```python
            elif gate.name == "SWAP":
                self._update_swap_gate(gate)
                continue
```
should be moved to later part, together with other `_update_*` methods.